### PR TITLE
Fix str of sum involving negative quantities.

### LINKED ISSUE #24022: `_print_Add` is fundamentally wrong
The `print_Add` in `printing/str.py` is making very strong and incorrect assumptions leading to incorrectly printed string (and c/latex/....) in a lot of cases (and this breaks our code generation in [Devito](https://github.com/devitocodes/devito) since incorrect sign leads to unstable kernels producing NaNs)

```
In [15]: from sympy import *

In [16]: a,b,c,d = symbols('a b c d')

In [17]: expr = Add(Add(-a, b, evaluate=False), Add(-c, d,  evaluate=False),  evaluate=False)

In [18]: expr
Out[18]: -(a + b) - (c + d) # this should be (-a + b) + (-c + d)
```

Basically, `_print_Add` always assume that any term `t` that is an Add that starts with a `-` is automatically a `-(t)` which is wrong since it would be a `Mul` .  The lines leading to the issues are lines 55-65  [Here](https://github.com/sympy/sympy/blob/45dcb6cae616eca10ca6df1a874d6aed25bc1fb4/sympy/printing/str.py#L55)

```            
            t = self._print(term)
            if t.startswith('-'):
                sign = "-"
                t = t[1:]
            else:
                sign = "+"
            if precedence(term) < PREC or isinstance(term, Add):
                l.extend([sign, "(%s)" % t])
            else:
                l.extend([sign, t])
```

That simply check if the string starts with `-` without any regards for the expression. so `-a +b` is printed as `- (a+b)` since it's an Add and starts with `-`.

### PR DESCRIPTION
**Pull request by Ishanned**
Fix unevaluated printing of sum involving negative quantities.
#### References to other Issues or PRs

fixes #24022 

#### Release Notes

* printing
   
  * Fix str of sum involving negative quantities.

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
It looks like you've fixed the problem with the backup files.

A test for the changed behaviour should be added though.

**Comment by smichr** [comment]
The suggested test in the OP of the issue passed with this change, so perhaps that could be added here.

**Comment by Ishanned** [comment]
@oscarbenjamin I have added the test, can you please check?

**Comment by Ishanned** [comment]
@smichr I have added the test, can you please check?

**Comment by asmeurer** [comment]
Please write more descriptive pull request titles. Just stating the file that you changed is not very helpful. Indicate what you actually fixed. The same applies to your commit messages.

File: sympy/printing/str.py [hunk 1]
```diff
@@ -54,7 +54,7 @@ def _print_Add(self, expr, order=None):
         l = []
         for term in terms:
             t = self._print(term)
-            if t.startswith('-'):
+            if t.startswith('-') and not term.is_Add:
```
Comment (mloubout) [inline sympy/printing/str.py:57]: Any reason to use `is_Add` here and `isinstance(term, Add)` below line 62? I think should both be `is_Add`

Comment (Ishanned) [reply sympy/printing/str.py:57]: @mloubout I have made the changes you asked for, please let me know if that does the work.

Comment (costika1234) [reply sympy/printing/str.py:57]: This change suggested by @smichr in https://github.com/sympy/sympy/issues/24022 appears to work fine, but more tests are needed to validate the fix.

Also, while we are at it, let's rename `PREC` to `prec` on line 53
```
PREC = precedence(expr)
```
and then on line 62:
```
if precedence(term) < PREC or term.is_Add:
```

File: sympy/printing/tests/test_str.py [hunk 2]
```diff
@@ -87,6 +87,7 @@ def test_Add():
     assert str(x - z*y**2*z*w) == "-w*y**2*z**2 + x"
     assert str(x - 1*y*x*y) == "-x*y**2 + x"
     assert str(sin(x).series(x, 0, 15)) == "x - x**3/6 + x**5/120 - x**7/5040 + x**9/362880 - x**11/39916800 + x**13/6227020800 + O(x**15)"
+    assert str(Add(Add(-w, x, evaluate=False), Add(-y, z,  evaluate=False),  evaluate=False)) == "(-w + x) + (-y + z)"
```
Comment (costika1234) [inline sympy/printing/tests/test_str.py:90]: Please add the following test cases that also validate the fix:
```
assert str(Add(Add(-x, -y, evaluate=False), -z, evaluate=False)) == "-z + (-x - y)"
assert str(Add(Add(Add(-x, -y, evaluate=False), -z, evaluate=False), -t, evaluate=False)) == "-t + (-z + (-x - y))"
```

**Review by costika1234 — CHANGES_REQUESTED** [review]
Please add the test cases I suggested and turn `PREC` into `prec`.

File: sympy/printing/tests/test_str.py (hunk 2 above)
Comment (costika1234) [reply sympy/printing/tests/test_str.py:90]: Everything looks good now but please make the title of your PR more descriptive as suggested above. You could say something like "Fix unevaluated printing of sum involving negative quantities." Also, add a release note, e.g. 

```
#### Release Notes
* printing
   
  * Fix str of sum involving negative quantities.
```

Comment (costika1234) [reply sympy/printing/tests/test_str.py:90]: That's okay. Note that the title of the PR is the one at the top of the page where it currently says "updated str.py #24030". You want to replace that with the description you added on the first line of your first message in this PR.

**Comment by costika1234** [comment]
@oscarbenjamin: everything in this PR looks good, albeit the title and commit could be better.

**Pull request merged**